getter.go

```diff
@@ -149,55 +244,68 @@ func (g *getter) worker() {
 	for c := range g.getCh {
 		g.retryGetChunk(c)
 	}
-
 }
 
-func (g *getter) retryGetChunk(c *chunk) {
+func (g *getter) retryGetChunk(c *Chunk) {
 	defer g.wg.Done()
 	var err error
-	c.b = <-g.sp.get
+	c.Bytes = <-g.sp.get
 	for i := 0; i < g.c.NTry; i++ {
 		time.Sleep(time.Duration(math.Exp2(float64(i))) * 100 * time.Millisecond) // exponential back-off
 		err = g.getChunk(c)
 		if err == nil {
 			return
 		}
-		logger.debugPrintf("error on attempt %d: retrying chunk: %v, error: %s", i, c.id, err)
+		logger.debugPrintf("error on attempt %d: retrying chunk: %v, error: %s", i, c.Id, err)
+		//If there is a re-used response in this chunk, it should be discarded since an error was returned
+		c.response = nil
 	}
 	g.err = err
+	c.Error = err
+	g.readCh <- c //expose error to the chunk handler
 	close(g.quit) // out of tries, ensure quit by closing channel
 }
 
-func (g *getter) getChunk(c *chunk) error {
-	// ensure buffer is empty
+func (g *getter) getChunk(c *Chunk) error {
+	var resp *http.Response
 
-	r, err := http.NewRequest("GET", g.url.String(), nil)
-	if err != nil {
-		return err
-	}
-	r.Header = c.header
-	g.b.Sign(r)
-	resp, err := g.c.Client.Do(r)
-	if err != nil {
-		return err
+	//The first chunk will have the initial response in it to re-use
+	if c.response == nil {
+
+		r, err := http.NewRequest("GET", c.url.String(), nil)
+		if err != nil {
+			return err
+		}
+		r.Header = c.header
+		g.b.Sign(r)
+		respTmp, err := g.c.Client.Do(r)
+		resp = respTmp //Necessary so 'resp' doesn't turn nil
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode != 206 {
+			return newRespError(resp)
+		}
+	} else {
+		resp = c.response
 	}
+
+	var err error
 	defer checkClose(resp.Body, &err)
-	if resp.StatusCode != 206 {
```

Comment:
this check is actually moved up to line 286 so it only happens when we make a new request. Response code is 206 when the request includes a "Range:" header, otherwise its 200